ambry-router/src/main/java/com.github.ambry.router/EncryptDecryptThread.java

```diff
@@ -0,0 +1,264 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Thread responsible for encrypting and decrypting blob content and per blob keys. Thread listens to a job queue
+ * for new jobs and processes it one by one.
+ */
+public class EncryptDecryptThread implements Runnable {
+  private final BlockingQueue<JobInfo> jobQueue;
+  private final CryptoService cryptoService;
+  private final KeyManagementService kms;
+  private final AtomicBoolean enabled;
+  private final Lock lock;
+  private CountDownLatch shutdownLatch;
+
+  private static final Logger logger = LoggerFactory.getLogger(EncryptDecryptThread.class);
+
+  /**
+   * Instantiates {@link EncryptDecryptThread} with the job queue, {@link CryptoService} and {@link KeyManagementService}
+   * @param jobQueue the {@link BlockingQueue} that contains the jobs
+   * @param cryptoService the {@link CryptoService} to use to encrypt or decrypt
+   * @param kms the {@link KeyManagementService} to fetch keys from
+   */
+  EncryptDecryptThread(BlockingQueue<JobInfo> jobQueue, CryptoService cryptoService, KeyManagementService kms) {
+    this.jobQueue = jobQueue;
+    this.cryptoService = cryptoService;
+    this.kms = kms;
+    enabled = new AtomicBoolean(false);
+    lock = new ReentrantLock();
+  }
+
+  @Override
+  public void run() {
+    enabled.set(true);
+    shutdownLatch = new CountDownLatch(1);
+    try {
+      while (enabled.get()) {
+        lock.lock();
+        try {
+          JobInfo jobInfo = jobQueue.poll(10, TimeUnit.MILLISECONDS);
+          if (jobInfo != null) {
+            jobInfo.doOperation(cryptoService, kms);
+          }
+        } catch (InterruptedException e) {
+          logger.error("InterruptedException thrown in EncryptDecryptThread");
+        } finally {
+          lock.unlock();
+        }
+      }
+    } finally {
+      shutdownLatch.countDown();
+    }
+  }
+
+  /**
+   * Shuts down the thread by closing all pending jobs
+   */
+  void close() {
+    lock.lock();
+    try {
+      if (enabled.compareAndSet(true, false)) {
+        logger.info("Closing EncryptDecryptThread");
+        closePendingJobs(new GeneralSecurityException("Shutting down EncryptDecrypt Thread"));
+      }
+    } finally {
+      lock.unlock();
+    }
+    if (shutdownLatch != null) {
+      try {
+        shutdownLatch.await(10, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        logger.error("InterruptedException thrown while waiting for EncryptDecryptThread to shutdown");
+      }
+    }
+  }
+
+  /**
+   * Closes all pending jobs with the given {@code e}
+   * @param gse the {@link GeneralSecurityException} that needs to be set while invoking callbacks for the pending jobs
+   */
+  private void closePendingJobs(GeneralSecurityException gse) {
```

Comment:
This doesn't seem to prevent people from enqueueing more jobs in the future, since the queue is passed in.

A way to fix this could be to keep the queue within this class and add an `enqueueJob()` method.

Reply:
Unfortunately, plan is up spin up multiple such worker thread to scale. So, we can't have one such API. But the router will ensure to shutdown all operations before shutting down the CryptoWorker. so this is taken care of